src/Microsoft.OData.Core/JsonLight/ODataJsonLightReader.cs

```diff
@@ -2256,6 +2259,117 @@ private void EndEntry()
+
+                            // Skip declared properties that are not null-able types.
+                            // Skip navigation properties (no navigation link means link value is null)
+                            // Skip properties that have been read (primitive or structural).
+                            if (!currentProperty.Type.IsNullable
+                                || currentProperty.PropertyKind == EdmPropertyKind.Navigation
+                                || resource.Properties.Any(p => p.Name.Equals(currentProperty.Name, StringComparison.Ordinal))
+                                || resourceState.NavigationPropertiesRead.Contains(currentProperty.Name))
+                            {
+                                continue;
+                            }
+
+                            RestoreNullODataProperty(currentProperty.Name, resourceState);
+                        }
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Restore the null property value in the resource.
+        /// </summary>
+        /// <param name="nullPropertyName">Name of the property to restore the null value.</param>
+        /// <param name="resourceState">The resource state.</param>
+        private static void RestoreNullODataProperty(string nullPropertyName, IODataJsonLightReaderResourceState resourceState)
+        {
+            // Add null value to omitted property declared in the type
+            ODataProperty property = new ODataProperty
+            {
+                Name = nullPropertyName,
+                Value = new ODataNullValue()
+            };
+
+            // Mark as processed, will throw if duplicate is detected.
+            resourceState.PropertyAndAnnotationCollector.MarkPropertyAsProcessed(property.Name);
+            ODataResourceBase resource = resourceState.Resource;
+
+            ReadOnlyEnumerable<ODataProperty> readonlyEnumerable =
+                resource.Properties as ReadOnlyEnumerable<ODataProperty>;
+            if (readonlyEnumerable != null)
+            {
+                // For non-entity resource, add the property underneath the read-only enumerable.
+                resource.Properties.ConcatToReadOnlyEnumerable("Properties", property);
+            }
+            else
+            {
+                // For entity resource, concatenate the property to original enumerable.
+                resource.Properties =
+                    resource.Properties.Concat(new List<ODataProperty>() { property });
+            }
+        }
+
+        /// <summary>
+        /// Get all properties defined by the EDM structural type and its base types.
+        /// </summary>
+        /// <param name="edmStructuredType">The EDM structural type.</param>
+        /// <returns>All the properties of this type.</returns>
+        private static IEnumerable<IEdmProperty> GetSelfAndBaseTypeProperties(IEdmStructuredType edmStructuredType)

```

Comment:
>GetSelfAndBaseTypeProperties [](start = 49, length = 28)

Note that you can use the Properties() extension method to get all declared and inherited properties for a structured type.

Reply:
OK. Thanks.

---
In reply to: [198061988](https://github.com/OData/odata.net/pull/1106#discussion_r198061988) [](ancestors = 198061988)